forc-pkg/src/manifest/mod.rs

```diff
@@ -322,13 +323,27 @@ impl DependencyDetails {
             branch,
             tag,
             rev,
+            version,
+            ipfs,
+            namespace,
             ..
         } = self;
 
         if git.is_none() && (branch.is_some() || tag.is_some() || rev.is_some()) {
             bail!("Details reserved for git sources used without a git field");
         }
 
+        if version.is_some() && git.is_some() {
+            bail!("Both version and git details provided for dependency");
+        }
+
+        if version.is_some() && ipfs.is_some() {
+            bail!("Both version and ipfs details provided for dependency");
+        }
+
+        if version.is_none() && namespace.is_some() {
+            bail!("Namespace can only be specified for sources with version");
+        }
```

Comment:
We can do this in the command definition using clap's `conflicts_with`

Reply:
oh nice one! I will take a look at that

Reply:
I came to this comment to apply but realized these values are coming from a TOML file not CLI so not so sure about how to use clap's conflict with groups to enforce this